### PR TITLE
Use the default_branch as the name of an initialized branch instead of defaulting to master

### PR DESCRIPTION
**Pull request by jamiemccrindle**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

Currently if you initialize a repo, it will always create a 'master' branch. If you specify a 'default_branch' that isn't master, setting the default branch will fail. This change means that the value of 'default_branch' is used as the branch name on initialization e.g. if you wanted the branch to be called 'refs/heads/main'. Note: this only changes the behaviour for 'Clean' initialization type. For uninitialized, there won't be a branch to set.

Issue Number: No issue

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [x] No

## Does this introduce a breaking change?

- [x] No

This does change the behavior in that 'default_branch' now works when initializing repos where previously it didn't. 

## Any relevant logs, error output, etc?

n/a

## Other information

No component changes.

### CODE REVIEW

**Comment by xuzhang3** [comment]
Hi @jamiemccrindle Thanks for opening this PR.  I try update the repository `default_branch` and I will get an error :
image
Because the branch I try set as default branch does not exist.

**Comment by jamiemccrindle** [comment]
Oh dang. I’ll test further. Can you send me the tf that broke?

**Comment by xuzhang3** [comment]
@jamiemccrindle 
Reproduce steps:
1. `terraform apply`
2. change `default_branch` to another branch  like `refs/heads/release`
3. `terraform apply`

TF config:
```hcl
resource "azuredevops_project" "project" {
  project_name       = "Sample Project"
  visibility         = "private"
  version_control    = "Git"
  work_item_template = "Agile"
}

resource "azuredevops_git_repository" "repo" {
  project_id = azuredevops_project.project.id
  name       = "Sample Empty Git Repository"
  initialization {
    init_type = "Clean"
  }
  default_branch="refs/heads/master"
}
```

**Comment by jamiemccrindle** [comment]
Thank you!

Testing now. I suspect that this may have broken before too i.e. the initalization only happens once and the default branch needs the branch to exist but will confirm.

**Comment by xuzhang3** [comment]
Set `default_branch` attribute `ForceNew: true,` might be a choice for this issue. A new repo will be created and destroy the old one .

**Comment by jamiemccrindle** [comment]
@xuzhang3, have confirmed that this would also break in trunk i.e. changing the default branch after creation to a branch that doesn't exist. If adding `force_new` it useful it might be better to add it in another PR.

**Comment by EliiseS** [comment]
@xuzhang3 @jamiemccrindle Could one of you create an issue for the bug you discovered? 😛

**Comment by jamiemccrindle** [comment]
hi @EliiseS, is it the one for a change of the `default_branch` setting failing on update if the branch doesn't exist? Or an issue for this enhancement?

**Comment by EliiseS** [comment]
The change of the default_branch setting failing on update if the branch doesn't exist @jamiemccrindle  :)

File: azuredevops/internal/service/git/resource_git_repository.go [hunk 1]
```diff
@@ -227,14 +227,18 @@ func createGitRepository(clients *client.AggregatedClient, repoName *string, pro
 	return createdRepository, nil
 }
 
-func initializeGitRepository(clients *client.AggregatedClient, repo *git.GitRepository) error {
+func initializeGitRepository(clients *client.AggregatedClient, repo *git.GitRepository, defaultBranch *string) error {
```
Comment (EliiseS) [inline azuredevops/internal/service/git/resource_git_repository.go:230]: Could you also create a test for this change?

Comment (jamiemccrindle) [reply azuredevops/internal/service/git/resource_git_repository.go:230]: sure thing! have added a test.

**Comment by jamiemccrindle** [comment]
re: raising an issue for the `default_branch` failing on update if the branch doesn't exist. It's a bit of a tricky one. I think it's reasonable if an update fails if the branch we're setting as the `default_branch` doesn't exist.

what's arguably more interesting is that the `default_branch` works when the repo is first created even if the branch doesn't exist.  this PR fixes it for the case where the initalization type is 'clean' but it's probably still wrong for uninitialized repos. Only problem is that I think that fixing it for uninitialized repos could be considered a breaking change.

**Comment by xuzhang3** [comment]
@jamiemccrindle If you set `initialization` with init_type: Clean`, the  default branch  `master` will be created.  `init_type: uninitialized `   won't create the default branch which means no branch exist for the new created repo.

**Comment by xuzhang3** [comment]
@jamiemccrindle LGTM, thanks for your contribute.

**Pull request merged**